java/ql/test/utils/model-generator/typebasedflow/CaptureTypeBasedSummaryModels.expected

```diff
@@ -0,0 +1,124 @@
+| p;MyFunction;true;apply;(Object,Object);;Argument[-1].SyntheticField[ArgType2];ReturnValue;value;generated |
```

Comment:
The naming of the synthetic fields could be a bit more precise. In particular they ought to refer to the type (when they aren't derived from a super-class/-interface), or reuse the field that the supertype indicates. I.e. a class `MyClass<T>` can use `SyntheticField[package.MyClass.T]`, a class `MySubclass<T2> extends MyClass<T2>` should use the same field as the superclass `SyntheticField[package.MyClass.T]`, and an extension of known collection-types should reuse those, such that `MyMultiMap<K,V> implements Map<K,List<V>>` uses `MapKey` and `MapValue.Element`. (The test `TypeBasedCollection<T> extends ArrayList<T>` already checks that to some degree).
Note that this will also ensure that methods on `A` and `B` will be compatible when we have `class A<T1,T2> extends B<T2>`, whereas the current approach fails to do so (equating the wrong fields and misses the equality that ought to be present).

Reply:
Given that a class can implement multiple interfaces, then we could have a conflict, so in that case the preference order should probably be something like "known collection-like super-interface" > "super-class" > "super-interface".

Reply:
Also, could we make this into some sort of inline test expectation? We don't necessarily need to use the inline expectation library - a more ad-hoc approach could be fine: I'm thinking perhaps collect all comments of the form `// MaD = ...` and then compare the query output with the set of those strings. We don't necessarily need to make use of the comment locations, which means that we could just write several expectation lines above the method that generates them.

Reply:
> The naming of the synthetic fields could be a bit more precise. In particular they ought to refer to the type (when they aren't derived from a super-class/-interface), or reuse the field that the supertype indicates. I.e. a class `MyClass<T>` can use `SyntheticField[package.MyClass.T]`, a class `MySubclass<T2> extends MyClass<T2>` should use the same field as the superclass `SyntheticField[package.MyClass.T]`, and an extension of known collection-types should reuse those, such that `MyMultiMap<K,V> implements Map<K,List<V>>` uses `MapKey` and `MapValue.Element`. (The test `TypeBasedCollection<T> extends ArrayList<T>` already checks that to some degree). Note that this will also ensure that methods on `A` and `B` will be compatible when we have `class A<T1,T2> extends B<T2>`, whereas the current approach fails to do so (equating the wrong fields and misses the equality that ought to be present).

You are absolutely right that the implementation could be improved and that the precision for more complicated inheritance is not ideal. The same is the case for C#, which also lacks exactly what you are describing here.
Do you prefer that we get this sorted as a first attempt before merging this?

Reply:
> Do you prefer that we get this sorted out as a first attempt before merging this?

I'll leave that to your preference. But I would like a refactor of the test, such that the models could be seen directly in the Java source next to the method.

Reply:
Something like
```
string expects() {
  exists(JavaDoc doc |
    doc.getChild(0).toString().regexpCature(" *MaD *= *\\(.*\\)", 1) = result
  )
}

...
```

Reply:
Yes, this is a great idea. I have included in it in this PR for the type based summaries, but maybe we should consider to do the same thing for both Java and C# for all model generator like tests.